### PR TITLE
improve error when outcome is missing

### LINKED ISSUE #1003: Better error message if outcome has 0 columns
Extracted from: https://stackoverflow.com/questions/77305226/why-am-i-receiving-the-error-object-not-found

If someone uses `step_dummy()` or some other ways in {recipes} to remove the outcome, then `parsnip:::check_outcome()` throws a wonderfully bad error 😄 

``` r
library(tidymodels)

rec <- recipe(Species ~ ., data = iris) %>%
  step_dummy(Species)

svm_spec <- 
  svm_rbf() %>% 
  set_engine("kernlab") %>% 
  set_mode("classification")

wflow <- workflow() %>% 
  add_model(svm_spec) %>% 
  add_recipe(rec)

fit(wflow, iris)
#> Error: object '.' not found
```

<sup>Created on 2023-10-16 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### PR DESCRIPTION
**Pull request by simonpcouch**
Closes #1003. :)

### CODE REVIEW

File: R/misc.R [hunk 1]
```diff
@@ -354,6 +354,16 @@ check_outcome <- function(y, spec) {
     return(invisible(NULL))
   }
 
+  has_no_outcome <- if (is.atomic(y)) {is.null(y)} else {length(y) == 0}
```
Comment (simonpcouch) [inline R/misc.R:357]: For `is.atomic(y)`, note `is.atomic(NULL)` is `TRUE`, and for `length(y) == 0`, `y` can either be a list or tibble.

File: R/misc.R [hunk 2]
```diff
@@ -354,6 +354,16 @@ check_outcome <- function(y, spec) {
...
+  if (isTRUE(has_no_outcome)) {
+    cli::cli_abort(
+      c("!" = "{.fun {class(spec)[1]}} was unable to find an outcome.",
+        "i" = "Ensure that you have specified an outcome column and that it \\
+               hasn't been removed in pre-processing."),
+      call = NULL
```
Comment (simonpcouch) [inline R/misc.R:363]: Plumbing the `call` here is a bit tricky. We _could_ pass this call all the way along from wherever it came from in parsnip (`fit()` or `fit_xy()`), but that's not necessarily the function called by the user (could be any function from tune, workflowsets, etc). Instead, I opt to call out the name of the model spec, which the user definitely wrote out at some point.

File: NEWS.md [hunk 3]
```diff
@@ -1,5 +1,7 @@
 # parsnip (development version)
 
+* Improved errors in cases where the outcome column is mis-specified.
```
Comment (EmilHvitfeldt) [inline NEWS.md:3]:
```suggestion
* Improved errors in cases where the outcome column is mis-specified. (#1003)
```

File: R/misc.R (hunk 2 above)
Comment (EmilHvitfeldt) [reply R/misc.R:363]: Yes this seems wise. Because there can be many reasons why an outcome didn't get passed to the model

**Review by EmilHvitfeldt — APPROVED** [review]

**Pull request merged**